File: app/src/main/java/alex/com/githubchecker/components/app/api/APIClient.java
package alex.com.githubchecker.components.app.api;

import java.io.IOException;
import java.util.List;

import alex.com.githubchecker.models.Diff;
import alex.com.githubchecker.models.api.PullRequest;
import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import timber.log.Timber;

/**
 * Created by Alex on 11/17/2017.
 */

public class APIClient {

    private GithubAPI _githubAPI;
    private OkHttpClient _client;

    public APIClient(OkHttpClient client, GithubAPI githubAPI) {
        _githubAPI = githubAPI;
        _client = client;
    }

    public Observable<List<PullRequest>> getPullRequests(String owner, String repo) {
        return _githubAPI.getPullRequests(owner, repo);
    }

    //We cant use a standard API call for this, so we have to manually get the data
    //From the URL provided
    public Observable<String> getDiffForPullRequest(PullRequest pullRequest) {
        Request.Builder req = new Request.Builder();
        req.url(pullRequest.getDiffUrl());
        PublishSubject<String> publishSubject = PublishSubject.create();

        _client.newCall(req.build()).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                Timber.e("getDiffForPullRequest FAILURE()");
                publishSubject.onError(e);
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                if (response.isSuccessful() && response.body() != null) {
                    publishSubject.onNext(response.body().string());
                }
            }
        });
        return publishSubject;
    }
}

File: app/src/main/java/alex/com/githubchecker/components/app/api/APIClient.kt
package alex.com.githubchecker.components.app.api

import alex.com.githubchecker.models.api.PullRequest
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject
import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import timber.log.Timber
import java.io.IOException

/**
 * Created by Alex on 11/17/2017.
 */

class APIClient(private val _client: OkHttpClient, private val _githubAPI: GithubAPI) {

    fun getPullRequests(owner: String, repo: String): Observable<List<PullRequest>> {
        return _githubAPI.getPullRequests(owner, repo)
    }

    //We cant use a standard API call for this, so we have to manually get the data
    //From the URL provided
    fun getDiffForPullRequest(pullRequest: PullRequest): Observable<String> {
        val req = Request.Builder()
        req.url(pullRequest.diff_url!!)
        val publishSubject = PublishSubject.create<String>()

        _client.newCall(req.build()).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                Timber.e("getDiffForPullRequest FAILURE()")
                publishSubject.onError(e)
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful && response.body != null) {
                    publishSubject.onNext(response.body!!.string())
                }
            }
        })
        return publishSubject
    }
}
